catalog/src/main/java/org/killbill/billing/catalog/DefaultCatalogService.java

```diff
@@ -49,20 +59,44 @@
     private boolean isInitialized;
 
     private final TenantInternalApi tenantInternalApi;
+    private final PriceOverride priceOverride;
 
     private final CatalogCache catalogCache;
     private final CacheInvalidationCallback cacheInvalidationCallback;
 
+    private final InternalCallContextFactory contextFactory;
+    private final CatalogPluginApi pluginCatalog;
+    private final Clock clock;
+
     @Inject
     public DefaultCatalogService(final CatalogConfig config,
                                  final TenantInternalApi tenantInternalApi,
                                  final CatalogCache catalogCache,
-                                 @Named(CatalogModule.CATALOG_INVALIDATION_CALLBACK) final CacheInvalidationCallback cacheInvalidationCallback) {
+                                 @Named(CatalogModule.CATALOG_INVALIDATION_CALLBACK) final CacheInvalidationCallback cacheInvalidationCallback,
+                                 final InternalCallContextFactory contextFactory,
+                                 final OSGIServiceRegistration<CatalogPluginApi> pluginRegistry,
+                                 final PriceOverride priceOverride,
+                                 final Clock clock
+                                ) {
+        super();
         this.config = config;
         this.catalogCache = catalogCache;
         this.cacheInvalidationCallback = cacheInvalidationCallback;
         this.tenantInternalApi = tenantInternalApi;
+        this.priceOverride = priceOverride;
         this.isInitialized = false;
+        this.contextFactory = contextFactory;
+        this.clock = clock;
+        this.pluginCatalog = selectCatalogPlugin(pluginRegistry);
```

Comment:
:warning: Don't cache it - plugins can come and go at runtime. You'll have to look at the registry each time (see `InvoicePluginDispatcher#getInvoicePlugins`).